packages/ai-jsx/src/core/conversation.tsx

```diff
@@ -0,0 +1,295 @@
+  return messages.map((message) => {
+    switch (message.role) {
+      case 'system':
+        return <SystemMessage>{message.content}</SystemMessage>;
+      case 'user':
+        return <UserMessage>{message.content}</UserMessage>;
+      case 'assistant':
+        return <AssistantMessage>{message.content}</AssistantMessage>;
+      case 'function':
+        return (
+          <FunctionCall name={message.function_call!.name!} args={JSON.parse(message.function_call!.arguments!)} />
+        );
+    }
+  });
+}
+
+/**
+ * Provide a function call to the LLM, for use within a {@link ChatCompletion}.
+ *
+ * The function call tells the model that a function was previously invoked by the model. See https://platform.openai.com/docs/guides/gpt/chat-completions-api for more detail.
+ * When the model returns a function call, @{link ChatCompletion} returns a @{link FunctionCall} component.
+ *
+ * @example
+ * ```tsx
+ *    <ChatCompletion>
+ *      <UserMessage>What is 258 * 322?</UserMessage>
+ *      <FunctionCall name="evaluate_math" args={expression: "258 * 322"} />
+ *      <FunctionResponse name="evaluate_math">83076</FunctionResponse>
+ *    </ChatCompletion>
+ *
+ *    ==> "That would be 83,076."
+ * ```
+ */
+export function FunctionCall({
+  name,
+  partial,
+  args,
+}: {
+  name: string;
+  partial?: boolean;
+  args: Record<string, string | number | boolean | null>;
+}) {
+  return `Call function ${name} with ${partial ? '(incomplete) ' : ''}${JSON.stringify(args)}`;
+}
+
+/**
+ * Renders to the output of a previous {@link FunctionCall} component, for use within a {@link ChatCompletion}.
+ *
+ * See https://platform.openai.com/docs/guides/gpt/chat-completions-api for more detail.
+ *
+ * @example
+ * ```tsx
+ *    <ChatCompletion>
+ *      <UserMessage>What is 258 * 322?</UserMessage>
+ *      <FunctionCall name="evaluate_math" args={expression: "258 * 322"} />
+ *      <FunctionResponse name="evaluate_math">83076</FunctionResponse>
+ *    </ChatCompletion>
+ *
+ *    ==> "That would be 83,076."
+ * ```
+ */
+export async function FunctionResponse(
+  { name, children }: { name: string; children: Node },
+  { render }: AI.ComponentContext
+) {
+  const output = await render(children);
+  return `function ${name} returns ${output}`;
+}
+
+interface ConversationMessageType<T, C extends AI.Component<any>> {
+  type: T;
+  element: AI.Element<AI.PropsOfComponent<C>>;
+}
+
+/**
+ * A type that represents a conversation message.
+ */
+export type ConversationMessage =
+  | ConversationMessageType<'user', typeof UserMessage>
+  | ConversationMessageType<'assistant', typeof AssistantMessage>
+  | ConversationMessageType<'system', typeof SystemMessage>
+  | ConversationMessageType<'functionCall', typeof FunctionCall>
+  | ConversationMessageType<'functionResponse', typeof FunctionResponse>;
+
+/** @hidden */
+export function isConversationalComponent(element: AI.Element<any>): boolean {
+  return (
+    element.tag === UserMessage ||
```

Comment:
This can also be done with something like `[UserMessage, /* ... */].includes(element.tag)`,